lib/parachain/candidate_validation_test.go

```diff
@@ -0,0 +1,131 @@
+			ErasureRoot:                 common.MustHexToHash("0xc07f658163e93c45a6f0288d229698f09c1252e41076f4caa71c8cbc12f118a1"), //nolint:lll
+			ParaHead:                    common.MustHexToHash("0x9a8a7107426ef873ab89fc8af390ec36bdb2f744a9ff71ad7f18a12d55a7f4f5"), //nolint:lll
+			ValidationCodeHash:          parachaintypes.ValidationCodeHash(validationCodeHashV),
+		},
+
+		CommitmentsHash: common.MustHexToHash("0xa54a8dce5fd2a27e3715f99e4241f674a48f4529f77949a4474f5b283b823535"),
+	}
+
+	payload, err := candidateReceipt.Descriptor.CreateSignaturePayload()
+	require.NoError(t, err)
+
+	signatureBytes, err := collatorKeypair.Sign(payload)
+	require.NoError(t, err)
+
+	signature := [sr25519.SignatureLength]byte{}
+	copy(signature[:], signatureBytes)
+
+	candidateReceipt.Descriptor.Signature = parachaintypes.CollatorSignature(signature)
+
+	return candidateReceipt, validationCode
+}
+
+func TestValidateFromChainState(t *testing.T) {
+
+	candidateReceipt, validationCode := createTestCandidateReceiptAndValidationCode(t)
+
+	bd, err := scale.Marshal(BlockDataInAdderParachain{
+		State: uint64(1),
+		Add:   uint64(1),
+	})
+	require.NoError(t, err)
+
+	pov := PoV{
+		BlockData: bd,
+	}
+
+	// NOTE: adder parachain internally compares postState with bd.State in it's validate_block,
+	// so following is necessary.
+	encodedState, err := scale.Marshal(uint64(1))
+	require.NoError(t, err)
+	postState, err := common.Keccak256(encodedState)
+	require.NoError(t, err)
+
+	hd, err := scale.Marshal(HeadDataInAdderParachain{
+		Number:     uint64(1),
+		ParentHash: common.MustHexToHash("0x0102030405060708090001020304050607080900010203040506070809000102"),
+		PostState:  postState,
+	})
+	require.NoError(t, err)
+
+	persistedValidationData := parachaintypes.PersistedValidationData{
+		ParentHead:             parachaintypes.HeadData{Data: hd},
+		RelayParentNumber:      uint32(1),
+		RelayParentStorageRoot: common.MustHexToHash("0x50c969706800c0e9c3c4565dc2babb25e4a73d1db0dee1bcf7745535a32e7ca1"),
+		MaxPovSize:             uint32(2048),
+	}
+
+	ctrl := gomock.NewController(t)
+
+	mockInstance := NewMockRuntimeInstance(ctrl)
+	mockInstance.EXPECT().
+		ParachainHostPersistedValidationData(
+			uint32(1000),
+			gomock.AssignableToTypeOf(parachaintypes.OccupiedCoreAssumption{})).
+		Return(&persistedValidationData, nil)
+	mockInstance.EXPECT().
+		ParachainHostValidationCode(uint32(1000), gomock.AssignableToTypeOf(parachaintypes.OccupiedCoreAssumption{})).
+		Return(&validationCode, nil)
+	mockInstance.EXPECT().
+		ParachainHostCheckValidationOutputs(uint32(1000), gomock.AssignableToTypeOf(parachaintypes.CandidateCommitments{})).
+		Return(true, nil)
+
+	mockPoVRequestor := NewMockPoVRequestor(ctrl)
+	mockPoVRequestor.EXPECT().
+		RequestPoV(common.MustHexToHash("0xe7df1126ac4b4f0fb1bc00367a12ec26ca7c51256735a5e11beecdc1e3eca274")).Return(pov)
+
+	_, _, _, err = ValidateFromChainState(mockInstance, mockPoVRequestor, candidateReceipt)
```

Comment:
Should we assert the returned values (*parachaintypes.CandidateCommitments, *parachaintypes.PersistedValidationData, and isValid)? or is that not necessary since we're essentially confirming the values returned from mocks?

Reply:
Added assertion for these. I can't test values in candidate commitment, since they are returned from runtime, but checking other things.